pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,709 @@
+
+    @step("Remove RPMs")
+    def remove_rpms(self, repo_map):
+        return self.delete_content(RpmUnit, repo_map, self._rpm_remove_crit)
+
+    @step("Remove ISOs")
+    def remove_files(self, repo_map):
+        return self.delete_content(FileUnit, repo_map, self._file_remove_crit)
+
+    @step("Remove modules")
+    def remove_modules(self, repo_map):
+        return self.delete_content(ModulemdUnit, repo_map, self._module_remove_crit)
+
+    def delete_content(self, unit_type, repo_map, criteria_fn):
+        out = []
+        # get repos
+        repos = self.search_repo(repo_map.keys())
+
+        # request removal
+        for repo in sorted(repos):
+            units = repo_map.get(repo.id)
+            criteria = self.unit_criteria(unit_type, criteria_fn(units))
+            f = repo.remove_content(criteria=criteria)
+            f = f_map(f, partial(ClearedRepo, repo=repo, content=units))
+            f = f_map(f, self.log_remove)
+            out.append(f)
+
+        return out
+
+    def _rpm_remove_crit(self, units):
+        part_crit = []
+        for unit in units:
+            part_crit.append(self._rpm_criteria(unit.filename, [unit.signing_key]))
+
+        return part_crit
+
+    def _file_remove_crit(self, units):
+        part_crit = []
+        for unit in units:
+            part_crit.append(self._file_criteria(unit.path))
+
+        return part_crit
+
+    def _module_remove_crit(self, units):
+        part_crit = []
+        for unit in units:
+            part_crit.append(self._module_criteria(unit.nsvca))
+
+        return part_crit
+
+    def search_repo(self, repo_ids):
+        return self.pulp_client.search_repository(Criteria.with_id(repo_ids)).result()
+
+    @step("Remove packages in advisory")
+    def remove_adv_items(self, repo_map):
+        out = []
+        # get repos from pulp
+        repos = self.search_repo(repo_map.keys())
+
+        for repo in repos:
+            items = repo_map.get(repo.id)
+            f = self.remove_adv_content(repo, items)
+            f = f_map(f, partial(ClearedRepo, repo=repo, content=items))
+            f = f_map(f, self.log_remove)
+            out.append(f)
+
+        return out
+
+    def remove_adv_content(self, repo, items):
+        crit = []
+        for item in items:
+            crit.append(
+                Criteria.and_(
+                    Criteria.with_field("filename", item.name),
+                    Criteria.with_field("signing_key", item.signing_key.lower()),
+                )
+            )
+
+        criteria = Criteria.and_(Criteria.with_unit_type(RpmUnit), Criteria.or_(*crit))
+
+        return repo.remove_content(criteria=criteria)
+
+    @step("Sort out files")
+    def separate_files(self, filenames):
+        rpms = []
+        modules = []
+        files = []
+        skipped = []
+        for f in filenames:
+            if f.endswith(".rpm"):
+                rpms.append(f)
+            elif f.endswith(".iso"):
```

Comment:
This looks wrong.  Nothing in here should be limited to removal of .iso files specifically.  "ISO" repositories in Pulp contain files not matching this pattern, so this would prevent the command from handling them.

I think the approach is just flawed actually.  You can't possibly know ahead of time if a certain file is going to be present in Pulp as an rpm unit, iso unit or something else just based on the filename.  For example: it's possible (maybe by accident) for someone to take an RPM file and upload it as a generic file (iso) into a FileRepository.  If that happened, then this command is the one we'd want to use to fix the situation and delete the file from the repo so it can be repushed to the correct repo.  But logic like "if filename ends in .rpm then we'll only look for it in RPM repositories" would prevent that.

If the user passed IDs of any FileRepository, I think we need to be searching for *all* listed filenames in that repository, even if 99%+ of the time we won't find RPMs there.

Reply:
I see what you mean. So it should look for all the requested files in the requested repos(file and yum repos only) and remove whatever is found in each repo (generating the criteria for removal based on the Unit type)?
I think this should be better with the available apis like repository.search_content rather than searching all the files in Pulp, verifying the requested repos against their repository_memberships and then removing the matched content from the repos.

Reply:
Yes, for example, it could be done by:

- if user requested delete of "some-file", then:
   - always do a search for FileUnit with path="some-file"
   - if "some-file" can be parsed as an RPM filename, then also do a search for RpmUnit using the parsed value
   - if "some-file" can be parsed as an NSVCA, then also do a search for ModulemdUnit using the parsed value
- when you've got all matching units, you can look at their repository_memberships, look at the repos passed in by the user and plan out what needs to be deleted.